### DISCUSSION TITLE
How to add caller's parent (containing class) to call hierarchy view?

### DISCUSSION THREAD

**Discussion by travkin79** [eclipse/lsp4e#978] [Q&A]
When viewing a call hierarchy In JDT and CDT, we're used to see the called methods, but also their parent classes. That's something we're missing in CDT LSP. It seems that's a feature that could be implemented in LSP4E. But I'm not sure how.

*For comparison:*

CDT's call hierarchy example:
![image](https://github.com/eclipse/lsp4e/assets/10887297/3d021332-1848-4148-a0f7-974ee9d633e3)

CDT LSP's call hierarchy example:
![image](https://github.com/eclipse/lsp4e/assets/10887297/dbfe9b02-f589-481c-88d3-66915b6b1b7d)

The call hierarchy feature is implemented in LSP4E. When debugging `org.eclipse.lsp4e.callhierarchy.CallHierarchyContentProvider#updateCallers(CallHierarchyViewTreeNode)` I didn't see any details about the calling method's parent / container in the language server response. It seems, the [`CallHierarchyItem`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#callHierarchyItem) does not provide any information on that, just the calling method's code range.

Is there a simpler way of retrieving the calling method's parent class than asking the language server for the [method's declaration](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_declaration) and then requesting the [document symbols ](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol) for a certain range of code?

**Comment by mickaelistria** [comment]
Can't your LS just set a more detailed name and/or set the `details` of the CallHierarchyItem ?

**Comment by travkin79** [comment]
Hi @mickaelistria,
Thanks for your answer and your suggestions. Especially using the `details` field in a `CallHierarchyItem` looks very promising.

[CDT LSP](https://github.com/eclipse-cdt/cdt-lsp) uses the [clangd](https://github.com/clangd/clangd) project and it seems, populating the details field with a file name, a containing class name, or method signature has already been discussed for clangd (see https://github.com/clangd/clangd/discussions/1940 and https://github.com/clangd/clangd/issues/1185). Unfortunately, it seems to be unlikely for these changes in clangd to be implemented any time soon.

I hoped for a more generic solution, but that would probably involve modifying the LSP standard. I hoped that I could adapt the call hierarchy view for CDT LSP based on already available features (or after minimal changes in LSP4E), but that seems a bit complicated and probably unnecessarily inefficient in terms of runtime.

**Comment by travkin79** [comment]
Hi all,

Good news: the language server implemented by clangd (vers. 20) [now also provides the called method's container](https://github.com/clangd/clangd/issues/1185#issuecomment-2689642955) (class and/or namespace) in the `detail` field.

Though, it looks somewhat odd in Eclipse (see screenshot). For example, there are some spaces missing (and the method name is contained twice in the labels).

![image](https://github.com/user-attachments/assets/24fdfd3d-1f8c-4471-ac6d-383880d0f4a1)

@mickaelistria and @rubenporras, do you know if any other language server (besides clangd) fills the `detail` field and if the labels look odd for other languages (language servers), too? I'd like to figure out what has to be fixed in LSP4E and what has to be fixed in clangd.

I didn't find any tests for the `CallHierarchyLabelProvider`. Are there any tests for checking the labels calculated for the `CallHierarchyItem`s? [According to the LSP4E source code the implementation is similar to that in VSCode](https://github.com/eclipse-lsp4e/lsp4e/blob/6e4750f1534b3b0d1700bc5367510119eeac6c5b/org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyLabelProvider.java#L65), but I couldn't find the corresponding VSCode source code for calculating the labels. @mickaelistria and @rubenporras, are you familiar with that? The feature was introduced with PR #513.

**Comment by travkin79** [comment]
I added issue #1231 and PR #1232.

I think, we can close this discussion now since the original problem is solved now.
